[SQLLINE-5] Use ServiceLoader to load drivers rather than forName

snuyanzin opened the pull request:
The PR implements usage of drivers mentioned by `META-INF/services/java.sql.Driver`

Also it allows to see errors in case of drivers loading which is impossible for current version.

Checked with latest drivers
- mysql
- oracle
- avatica
- calcite
- phoenix

fixes #5

arina-ielchiieva commented on src/main/java/sqlline/Application.java:
```diff
@@ -200,6 +201,7 @@ public String getVersion() {
    *
    * @see #DEFAULT_DRIVERS
    */
+  @Deprecated
```
@snuyanzin how list of known drivers can be overridden without using depicted method?

snuyanzin replied:
Could you please clarify what you mean?

Before this PR known drivers used to filter classes from classpath and to load drivers if they are matched to known drivers.
After PR changes java service provider functionality [1] will be used, i.e. java could go through jars looking for META-INF/services/java.sql.Driver and if success then load the driver.
So each driver which has such file inside its jar becomes known. It will allow to add more drivers in classpath and load them automatically without code changes

[1] https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html

arina-ielchiieva replied:
For Drill we were overriding all known drivers to only one - "org.apache.drill.jdbc.Driver" since Drill SqlLine is intended only for Apache Drill.

snuyanzin replied:
Looks like for your case it is required to have limit which could be done via _Application#initDrivers_. What about logic like if _Application#initDrivers_ is empty then any driver found via servicelocator is allowed. If _Application#initDrivers_ is not empty then only drivers which are present in both _Application#initDrivers_ and servicelocator are allowed.

However the name _Application#initDrivers_ becomes not matching to what it started to do. So it makes sense to add a new method with more suitable name and leave it deprecated and made call to the new one as it is public.

arina-ielchiieva replied:
Sounds good, just indicate in Javadoc which new method should be used.

julianhyde commented:
Don't represent "allow any driver" with empty collection. That causes a discontinuity going from 0 to 1 element. Use the null collection.

How about using a List rather than a Collection? Some applications that have a strong preference of one driver over another, so order might be important.

Please continue to add separate commits rather than squashing. I have already started reviewing the first commit, so separate follow-up commits will be easier for me to merge.

snuyanzin commented:
Thank you for your comment.
Changed Collection to List and started to use null collection to represent "allow any driver".
Added several more tests to check the work of limitation

julianhyde commented:
The two most recent commits have @snuyanzin's email address but use the name "Ekaterina". Do I need to preserve that authorship information or can I squash the commits?

snuyanzin commented:
it can be squashed
sorry..
looks like I forgot to change git username from default while working on my wife's laptop

julianhyde commented:
Merged as b053ffe, fixes #5.
